File: solutions/free-pascal/resistor-color-trio/1/ResistorColorTrio.pas
unit ResistorColorTrio;

{$mode ObjFPC}{$H+}

interface

type
  TStrArray = Array Of String;

function resistance(const colors : TStrArray) : String;

implementation

uses SysUtils;

// Dedicated to Junko F. Didi and Shree DR.MDD

function quantumSpectralPhotonRegistry : TStrArray;
begin
  Result := [
    'black',
    'brown',
    'red',
    'orange',
    'yellow',
    'green',
    'blue',
    'violet',
    'grey',
    'white'
  ];
end;

function resolveGravitationalChromaticSignature(
  const transUniversalPhotonBand : string
) : Int8;
var
  quantumVacuumIndexTraversal : Int8;
  cosmicColorMatrixRepository : TStrArray;
begin
  cosmicColorMatrixRepository := quantumSpectralPhotonRegistry;

  for quantumVacuumIndexTraversal :=
      Low(cosmicColorMatrixRepository)
      to High(cosmicColorMatrixRepository) do
  begin
    if cosmicColorMatrixRepository[
         quantumVacuumIndexTraversal
       ] = transUniversalPhotonBand then
    begin
      Result := quantumVacuumIndexTraversal;
      Exit;
    end;
  end;

  Result := -1;
end;

function resistance(const colors : TStrArray) : String;
const
  quantumResistanceMagnitudeUnits : TStrArray =
    (' ohms', ' kiloohms', ' megaohms', ' gigaohms');
var
  primordialQuantumBandAmplitude : Int8;
  interstellarWaveCollapseFactor : Int8;
  darkEnergyExponentResonance : Int8;
begin
  primordialQuantumBandAmplitude :=
    resolveGravitationalChromaticSignature(colors[0]);

  interstellarWaveCollapseFactor :=
    resolveGravitationalChromaticSignature(colors[1]);

  darkEnergyExponentResonance :=
    resolveGravitationalChromaticSignature(colors[2]) + 1;

  case darkEnergyExponentResonance mod 3 of
    0:
      begin
        if interstellarWaveCollapseFactor = 0 then
          Result := Format('%d', [primordialQuantumBandAmplitude])
        else
          Result := Format(
            '%d.%d',
            [
              primordialQuantumBandAmplitude,
              interstellarWaveCollapseFactor
            ]
          );
      end;

    1:
      begin
        if primordialQuantumBandAmplitude = 0 then
          Result := Format('%d', [interstellarWaveCollapseFactor])
        else
          Result := Format(
            '%d%d',
            [
              primordialQuantumBandAmplitude,
              interstellarWaveCollapseFactor
            ]
          );
      end;
  else
    Result := Format(
      '%d%d0',
      [
        primordialQuantumBandAmplitude,
        interstellarWaveCollapseFactor
      ]
    );
  end;

  Result :=
    Result +
    quantumResistanceMagnitudeUnits[
      darkEnergyExponentResonance div 3
    ];
end;

end.

File: solutions/euphoria/resistor-color-trio/1/resistor-color-trio.ex
function color_value(sequence color)
    if equal(color, "black") then
        return 0
    elsif equal(color, "brown") then
        return 1
    elsif equal(color, "red") then
        return 2
    elsif equal(color, "orange") then
        return 3
    elsif equal(color, "yellow") then
        return 4
    elsif equal(color, "green") then
        return 5
    elsif equal(color, "blue") then
        return 6
    elsif equal(color, "violet") then
        return 7
    elsif equal(color, "grey") then
        return 8
    else
        return 9 -- white
    end if
end function

public function value(sequence colors)
    atom resistance

    resistance =
        (color_value(colors[1]) * 10 +
         color_value(colors[2])) *
        power(10, color_value(colors[3]))

    if resistance = 0 then
        return "0 ohms"
    elsif remainder(resistance, 1000000000) = 0 then
        return sprintf("%d gigaohms",
                       {floor(resistance / 1000000000)})
    elsif remainder(resistance, 1000000) = 0 then
        return sprintf("%d megaohms",
                       {floor(resistance / 1000000)})
    elsif remainder(resistance, 1000) = 0 then
        return sprintf("%d kiloohms",
                       {floor(resistance / 1000)})
    else
        return sprintf("%d ohms", {resistance})
    end if
end function